lib/elixir/test/elixir/code_normalizer/general_test.exs

```diff
@@ -0,0 +1,616 @@
+
+      ast = quote(do: <<69 - 4::bits-size(8 - 4)-unit(1), 65>>)
+      assert quoted_to_string(ast) == "<<69 - 4::bits-size(8 - 4)-unit(1), 65>>"
+
+      ast = quote(do: <<(<<65>>), 65>>)
+      assert quoted_to_string(ast) == "<<(<<65>>), 65>>"
+
+      ast = quote(do: <<65, (<<65>>)>>)
+      assert quoted_to_string(ast) == "<<65, (<<65>>)>>"
+
+      ast = quote(do: for(<<(a::4 <- <<1, 2>>)>>, do: a))
+      assert quoted_to_string(ast) == "for <<(a::4 <- <<1, 2>>)>> do\n  a\nend"
+    end
+
+    test "charlist" do
+      assert quoted_to_string(quote(do: [])) == "[]"
+      assert quoted_to_string(quote(do: 'abc')) == "'abc'"
+    end
+
+    test "string" do
+      assert quoted_to_string(quote(do: "")) == ~S/""/
+      assert quoted_to_string(quote(do: "abc")) == ~S/"abc"/
+      assert quoted_to_string(quote(do: "#{"abc"}")) == ~S/"#{"abc"}"/
+    end
+
+    test "last arg keyword list" do
+      assert quoted_to_string(quote(do: foo([]))) == "foo([])"
+      assert quoted_to_string(quote(do: foo(x: y))) == "foo(x: y)"
+      assert quoted_to_string(quote(do: foo(x: 1 + 2))) == "foo(x: 1 + 2)"
+      assert quoted_to_string(quote(do: foo(x: y, p: q))) == "foo(x: y, p: q)"
+      assert quoted_to_string(quote(do: foo(a, x: y, p: q))) == "foo(a, x: y, p: q)"
+
+      assert quoted_to_string(quote(do: {[]})) == "{[]}"
+      assert quoted_to_string(quote(do: {[a: b]})) == "{[a: b]}"
+      assert quoted_to_string(quote(do: {x, a: b})) == "{x, [a: b]}"
+      assert quoted_to_string(quote(do: foo(else: a))) == "foo(else: a)"
+      assert quoted_to_string(quote(do: foo(catch: a))) == "foo(catch: a)"
+    end
+  end
+
+  describe "quoted_to_algebra/2 escapes" do
+    test "strings with slash escapes" do
+      assert quoted_to_string(quote(do: "\a\b\d\e\f\n\r\t\v"), escape: false) ==
+               ~s/"\a\b\d\e\f\n\r\t\v"/
+
+      assert quoted_to_string(quote(do: "\a\b\d\e\f\n\r\t\v")) ==
+               ~s/"\\a\\b\\d\\e\\f\\n\\r\\t\\v"/
+    end
+
+    test "strings with non printable characters" do
+      assert quoted_to_string(quote(do: "\x00\x01\x10"), escape: false) == ~s/"\x00\x01\x10"/
+      assert quoted_to_string(quote(do: "\x00\x01\x10")) == ~s/"<<0, 1, 16>>"/
+    end
+
+    test "charlists with slash escapes" do
+      assert quoted_to_string(quote(do: '\a\b\e\n\r\t\v'), escape: false) ==
+               ~s/'\a\b\e\n\r\t\v'/
+
+      assert quoted_to_string(quote(do: '\a\b\e\n\r\t\v')) ==
+               ~s/'\\a\\b\\e\\n\\r\\t\\v'/
+    end
+
+    test "charlists with non printable characters" do
+      assert quoted_to_string(quote(do: '\x00\x01\x10'), escape: false) == ~S/[0, 1, 16]/
+      assert quoted_to_string(quote(do: '\x00\x01\x10')) == ~S/[0, 1, 16]/
+    end
+
+    test "atoms" do
+      assert quoted_to_string(quote(do: :"a\nb\tc"), escape: false) == ~s/:"a\nb\tc"/
+      assert quoted_to_string(quote(do: :"a\nb\tc")) == ~S/:"a\nb\tc"/
+    end
+
+    test "atoms with interpolations" do
```

Comment:
We also need a test of atoms with non printable characters (they are kept as strings). :)

Reply:
Added!